packages/exhibitor/extra/bootstrap_exhibitor_tls.py

```diff
@@ -0,0 +1,213 @@
+#!/opt/mesosphere/bin/python
+import os
+import socket
+import subprocess
+import sys
+
+from pathlib import Path
+from urllib.parse import urlparse
+
+
+TLS_ARTIFACT_LOCATION = '/var/lib/dcos/exhibitor-tls-artifacts'
+CSR_SERVICE_CERT_PATH = '/tmp/root-cert.pem'
+PRESHAREDKEY_LOCATION = '/root/.dcos-bootstrap-ca-psk'
+EXHIBITOR_TLS_TMP_DIR = '/var/lib/dcos/exhibitor/.pki'
+BOOTSTRAP_CA_BINARY = '/opt/mesosphere/bin/dcos-bootstrap-ca'
+
+
+def invoke_detect_ip():
+    try:
+        ip = subprocess.check_output(
+            ['/opt/mesosphere/bin/detect_ip']).strip().decode('utf-8')
+    except subprocess.CalledProcessError as e:
+        print("check_output exited with {}".format(e))
+        sys.exit(1)
+    try:
+        socket.inet_aton(ip)
+        return ip
+    except socket.error as e:
+        print(
+            "inet_aton exited with {}. {} is not a valid IPv4 address".format(e, ip))
+        sys.exit(1)
+
+
+def get_ca_url(exhibitor_bootstrap_ca_url, bootstrap_url):
+    if exhibitor_bootstrap_ca_url:
+        print('Using `exhibitor_bootstrap_ca_url` config parameter.')
+        return exhibitor_bootstrap_ca_url
+    else:
+        print('Inferring `exhibitor_bootstrap_ca_url` from `bootstrap_url`.')
+        result = urlparse(bootstrap_url)
+
+        if result.scheme == 'http':
+            netloc = result.netloc.split(':', 1)  # strip port
+            return 'https://{}:443'.format(netloc[0])
+        elif result.scheme == 'file':
+            print('bootstrap url references a local file')
+        else:
+            print('bootstrap url is using an unsupported scheme: {}'.format(result.scheme))
+        return ""
+
+
+def test_connection(ca_url):
+    s = socket.socket()
+    s.settimeout(5)
```

Comment:
is this the connect() or the recv() timeout, or both?

Reply:
I mean, does this *really* set the connect() timeout, which is what we seem to care about here to not hang forever...

Reply:
We're good:

> The connect() operation is also subject to the timeout setting, and in general it is recommended to call settimeout() before calling connect()

https://docs.python.org/3/library/socket.html#timeouts-and-the-connect-method